compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/MainMethodGenerationLowering.kt

```diff
@@ -0,0 +1,92 @@
+import org.jetbrains.kotlin.ir.builders.irBlockBody
+import org.jetbrains.kotlin.ir.builders.irCall
+import org.jetbrains.kotlin.ir.builders.irReturn
+import org.jetbrains.kotlin.ir.declarations.IrClass
+import org.jetbrains.kotlin.ir.declarations.IrDeclarationOrigin
+import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
+import org.jetbrains.kotlin.ir.expressions.impl.IrConstImpl
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.ir.util.getAnnotation
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.resolve.DescriptorUtils
+import org.jetbrains.kotlin.types.Variance
+
+internal class MainMethodGenerationLowering(val context: JvmBackendContext) : ClassLoweringPass {
+
+    override fun lower(irClass: IrClass) {
+        if (!context.configuration.languageVersionSettings.supportsFeature(LanguageFeature.ExtendedMainConvention)) return
+        if (irClass.origin != IrDeclarationOrigin.FILE_CLASS) return
+
+        val parameterlessMain = irClass.functions.find { it.isParameterlessMainMethod() } ?: return
+
+        if (irClass.functions.any { it.isMainMethod() }) return
+
+        generateMainMethod(irClass, parameterlessMain)
+    }
+
+    private fun generateMainMethod(irClass: IrClass, parameterlessMain: IrSimpleFunction) {
+        irClass.addFunction {
+            name = Name.identifier("main")
+            visibility = Visibilities.PUBLIC
+            returnType = context.irBuiltIns.unitType
+            modality = Modality.FINAL
+            origin = JvmLoweredDeclarationOrigin.GENERATED_MAIN_FOR_PARAMETERLESS_MAIN_METHOD
+        }.apply {
+            addValueParameter {
+                name = Name.identifier("args")
+                type = context.irBuiltIns.arrayClass.typeWith(context.irBuiltIns.stringType)
+            }
+            body = context.createIrBuilder(this.symbol).irBlockBody {
+                +irReturn(this.irCall(parameterlessMain))
+            }
+        }
+    }
+
+    private fun IrSimpleFunction.isParameterlessMainMethod(): Boolean =
+        typeParameters.isEmpty() && valueParameters.isEmpty() && returnType.isUnit() && name.asString() == "main"
+
+
+    private fun IrSimpleFunction.isMainMethod(): Boolean {
+        if (!hasJvmNameMain) return false
+        if (!returnType.isUnit()) return false
+
+        val parameter = allParameters.singleOrNull() ?: return false
+        if (!parameter.type.isArray() && !parameter.type.isNullableArray()) return false
+
+        val argType = (parameter.type as IrSimpleType).arguments.first()
+        return when (argType) {
+            is IrTypeProjection -> {
+                (argType.variance != Variance.IN_VARIANCE) && argType.type.isStringClassType()
+            }
+            else -> false
+        }
+    }
+
+
+    private val IrSimpleFunction.hasJvmNameMain: Boolean
+        get() {
+            val jvmAnnotatedName =
```

Comment:
Minor, but this would probably be simpler:
```
val jvmName = (getAnnotation(DescriptorUtils.JVM_NAME)?.getValueArgument(0) as? IrConst<*>)?.value
        ?: name.asString()
return jvmName == "main"
```
Maybe even extracting and reusing `getJvmNameFromAnnotation` from `MethodSignatureMapper` would make sense

Reply:
Hah, yeah, that'd make sense. What's a good place to lift that out to? org.jetbrains.kotlin.backend.common.ir.IrUtils?

Reply:
And agreed on your review comment.

Reply:
> What's a good place to lift that out to? 

Something JVM-specific, e.g. `compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/ir/IrUtils.kt`

Reply:
Easy :) Done.